src/Common/src/TypeSystem/Interop/InteropStateManager.cs

```diff
@@ -0,0 +1,205 @@
+        //      1. Managed to Native Thunk: For forward marshalling
+        //      2. Native to Managed Thunk: For reverse marshalling
+        //      3. Cleanup Thunk: for cleaning up any allocated resources
+        //
+        /// <summary>
+        /// Generates a Native struct type which imitates the managed struct
+        /// </summary>
+        internal NativeStructType GetStructMarshallingNativeType(TypeDesc managedType)
+        {
+            if (managedType is ByRefType)
+            {
+                managedType = ((ByRefType)managedType).GetParameterType();
+            }
+
+            Debug.Assert(managedType is MetadataType);
+
+            var typeKey = new NativeStructTypeHashtableKey((MetadataType)managedType, _generatedAssembly);
+            return _nativeStructHashtable.GetOrCreateValue(typeKey);
+        }
+
+        /// <summary>
+        ///  Generates a thunk to marshal the fields of the struct from managed to native
+        /// </summary>
+        internal MethodDesc GetStructMarshallingManagedToNativeThunk(TypeDesc managedType)
+        {
+            if (managedType is ByRefType)
+            {
+                managedType = ((ByRefType)managedType).GetParameterType();
+            }
+
+            Debug.Assert(managedType is MetadataType);
+
+            var methodKey = new StructMarshallingThunkKey((MetadataType)managedType, _generatedAssembly.GetGlobalModuleType(), StructMarshallingThunkType.ManagedToNative);
+            return _structMarshallingThunkHashtable.GetOrCreateValue(methodKey);
+        }
+
+        /// <summary>
+        ///  Generates a thunk to marshal the fields of the struct from native to managed
+        /// </summary>
+        internal MethodDesc GetStructMarshallingNativeToManagedThunk(TypeDesc managedType)
+        {
+            if (managedType is ByRefType)
+            {
+                managedType = ((ByRefType)managedType).GetParameterType();
+            }
+
+            Debug.Assert(managedType is MetadataType);
+
+
+            var methodKey = new StructMarshallingThunkKey((MetadataType)managedType, _generatedAssembly.GetGlobalModuleType(), StructMarshallingThunkType.NativeToManage);
+            return _structMarshallingThunkHashtable.GetOrCreateValue(methodKey);
+        }
+
+        /// <summary>
+        ///  Generates a thunk to cleanup any allocated resources during marshalling
+        /// </summary>
+        internal MethodDesc GetStructMarshallingCleanupThunk(TypeDesc managedType)
+        {
+            if (managedType is ByRefType)
+            {
+                managedType = ((ByRefType)managedType).GetParameterType();
+            }
+
+            Debug.Assert(managedType is MetadataType);
+
+
+            var methodKey = new StructMarshallingThunkKey((MetadataType)managedType, _generatedAssembly.GetGlobalModuleType(), StructMarshallingThunkType.Cleanup);
+            return _structMarshallingThunkHashtable.GetOrCreateValue(methodKey);
```

Comment:
What happens if the struct does not require cleanup (such as 1-byte bool to 4 -byte BOOL)?